### PR TITLE
nephio-in-docker (nind)

### PR DESCRIPTION
**Pull request by danaschwanden**
nephio-in-docker (nind) packages the nephio-ansible-install into a [Docker](https://docs.docker.com/get-docker/) container following the setup used in the [nephio ONE summit 2022 workshop](https://github.com/nephio-project/one-summit-22-workshop).
In addition nind also runs a [VS Code OSS](https://github.com/microsoft/vscode) workstation instance providing a browser based development and test environment.

##### Core changes for new NIND feature
- nephio-ansible-install/Dockerfile
- nephio-ansible-install/nind/150_install-nephio.sh
- nephio-ansible-install/nind/nephio.yaml
- nephio-ansible-install/roles/cluster/deploy/tasks/cluster_files.yaml (extra port export for nephio webui ingress)
- nephio-ansible-install/roles/nephio/deploy/tasks/mgmtcluster_files.yaml (ingress for nephio webui)
- nephio-ansible-install/roles/nephio/deploy/templates/nephio-ingress.j2 (ingress for nephio webui)
- nephio-ansible-install/roles/gitea/create/tasks/main.yaml (ports on all interfaces to make them available for remote tunnelling)
- nephio-ansible-install/NIND.md
- nephio-ansible-install/diagrams/gitea-repos.png
- nephio-ansible-install/diagrams/nephio-webui.png
- nephio-ansible-install/diagrams/workstation.png
- nephio-ansible-install/roles/repo-gitea/create/tasks/main.yaml (pause to wait for gitea API to become available and avoid glitch)

##### Cosmetic changes to documentation
- nephio-ansible-install/README.md

###### Changes to TF code (run and tested the instance creation after changes)
- infra/compute_instances.tf
- infra/outputs.tf

###### Indentation/leading/trailing spaces removed
- ansible_kind/kind_setup.yaml
- nephio-ansible-install/playbooks/configure-nephio.yaml
- nephio-ansible-install/playbooks/create-gitea-repos.yaml
- nephio-ansible-install/playbooks/create-gitea.yaml
- nephio-ansible-install/playbooks/create-repos.yaml
- nephio-ansible-install/playbooks/delete-gitea-repos.yaml
- nephio-ansible-install/playbooks/delete-gitea.yaml
- nephio-ansible-install/playbooks/delete-repos.yaml
- nephio-ansible-install/playbooks/deploy-clusters.yaml
- nephio-ansible-install/playbooks/destroy-clusters.yaml
- nephio-ansible-install/playbooks/install-prereq.yaml
- nephio-ansible-install/playbooks/ping.yaml
- nephio-ansible-install/roles/clab/tasks/main.yaml
- nephio-ansible-install/roles/cluster/deploy/templates/cluster-config.yaml.j2
- nephio-ansible-install/roles/cluster/destroy/tasks/main.yaml
- nephio-ansible-install/roles/multus/tasks/cluster_files.yaml
- nephio-ansible-install/roles/multus/tasks/main.yaml
- nephio-ansible-install/roles/repo-gitea/delete/tasks/main.yaml

###### Double quotes added
- ansible_kind/scripts/multiclusterkind/clean.sh

### CODE REVIEW

**Comment by vjayaramrh** [comment]
@danaschwanden Would be great if you pushed a PR just with changes related to NIND only, Thanks. It would make it easier to review.

**Comment by danaschwanden** [comment]
> @danaschwanden Would be great if you pushed a PR just with changes related to NIND only, Thanks. It would make it easier to review.

Hi @vjayaramrh , I had to change the non NIND files to make the linter happy. If I only include the NIND related changes then the linter spits the dummy. Would you still like me to reverse those changes?

**Comment by vishwanathj** [comment]
> > @danaschwanden Would be great if you pushed a PR just with changes related to NIND only, Thanks. It would make it easier to review.
> 
> Hi @vjayaramrh , I had to change the non NIND files to make the linter happy. If I only include the NIND related changes then the linter spits the dummy. Would you still like me to reverse those changes?

@danaschwanden Thanks for clarifying. Alternate Suggestion (if it makes sense) -> Do you mind calling out in the PR body the list of files specific to NIND feature that needs review and add a note that the other files have been modified to pass the linter and have no actual content changes. Thanks

**Comment by danaschwanden** [comment]
> @danaschwanden Thanks for clarifying. Alternate Suggestion (if it makes sense) -> Do you mind calling out in the PR body the list of files specific to NIND feature that needs review and add a note that the other files have been modified to pass the linter and have no actual content changes. Thanks

Thanks a lot @vjayaramrh , had a quick exchange with @s3wong on slack earlier today and shared the following high level assessment with him:
- 4 out of the 35 files are outside the NIND context (not tested by me so far) and updated to keep the linter happy
- 31 files are related to NIND (tested implicitly by running NIND), of which 8 are net new (incl Dockerfile and pictures, etc)
- Most of the linter driven changes are removing whitespaces in curly brackets which the linter seems to particularly dislike

Long story short: The 4 files outside the NIND context boil down to the following:
- 2 TF files (infra/compute_instances.tf and infra/output.tf) would benefit from some testing. Reckon running the TF code will quickly tell whether it is all good or something has been broken.
- The other two files (from the ansible_kind folder) appear to be fairly benign changes as they are mostly removing spaces.

Can add that to the [first comment](https://github.com/nephio-project/one-summit-22-workshop/pull/113#issue-1516684143) above to make it more visible. Please lemme know if you think this would be helpful?

**Comment by henderiw** [comment]
I tested the complete flow with Dan's files and did some further changes, but I believe they are extensions from what Dan did.
In general I like the nind approach, as it make the installation portable by using the container environment.

I tested it on my laptop + my own VM in the cloud and in the google workstation.

File: nephio-ansible-install/NIND.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+# 00_prereq.out, 01_gitea.out, 02_repos.out, 03_clusters.out and 04_nephio.out
+# when the last step starts executing follow the progress by tailing its log file.
+tail -f 04_nephio.out
+
+# once the installtion completes you will see a line like the one below.
+PLAY RECAP *********************************************************************
+localhost                  : ok=22   changed=18   unreachable=0    failed=0    skipped=9    rescued=0    ignored=0
+```
+
+You can build and run nind on any laptop or workstaton that packs-the-punch to let nephio fly.
```
Comment (vishwanathj) [inline nephio-ansible-install/NIND.md:37]: For the workshop VM, we recommended a VM with 8 vcpus, 32 GB RAM and 50 GB disk space. I am guessing that the NIND container is to be run on a compute machine with above requirements, right?

Comment (vjayaramrh) [inline nephio-ansible-install/NIND.md:37]: For the workshop VM, we recommended a VM with 8 vcpus, 32 GB RAM and 50 GB disk space. I am guessing that the NIND container is to be run on a compute machine with above requirements, right?

File: nephio-ansible-install/NIND.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+nephio-in-docker (nind) packages the nephio-ansible-install into a [Docker](https://docs.docker.com/get-docker/) container following the setup used in the [nephio ONE summit 2022 workshop](https://github.com/nephio-project/one-summit-22-workshop).
+In addition nind also runs a [VS Code OSS](https://github.com/microsoft/vscode) workstation instance providing a browser based development and test enviroment.
+
+To build and run nephio-in-docker execute the commands listed below:
+
+```bash
+git clone https://github.com/nephio-project/one-summit-22-workshop.git
+cd one-summit-22-workshop/nephio-ansible-install
+docker build -t nind .
```
Comment (vjayaramrh) [inline nephio-ansible-install/NIND.md:13]: Would be great if we can provide instructions for podman as well. 
If it cannot be done in this PR, it might be good to open an issue (somebody else can open a PR and provide podman instructions) to track this.
Thoughts?

**Comment by vjayaramrh** [comment]
> > @danaschwanden Thanks for clarifying. Alternate Suggestion (if it makes sense) -> Do you mind calling out in the PR body the list of files specific to NIND feature that needs review and add a note that the other files have been modified to pass the linter and have no actual content changes. Thanks
> 
> Thanks a lot @vjayaramrh , had a quick exchange with @s3wong on slack earlier today and shared the following high level assessment with him:
> 
>     * 4 out of the 35 files are outside the NIND context (not tested by me so far) and updated to keep the linter happy
> 
>     * 31 files are related to NIND (tested implicitly by running NIND), of which 8 are net new (incl Dockerfile and pictures, etc)
> 
>     * Most of the linter driven changes are removing whitespaces in curly brackets which the linter seems to particularly dislike
> 
> 
> Long story short: The 4 files outside the NIND context boil down to the following:
> 
>     * 2 TF files (infra/compute_instances.tf and infra/output.tf) would benefit from some testing. Reckon running the TF code will quickly tell whether it is all good or something has been broken.
> 
>     * The other two files (from the ansible_kind folder) appear to be fairly benign changes as they are mostly removing spaces.
> 
> 
> Can add that to the [first comment](https://github.com/nephio-project/one-summit-22-workshop/pull/113#issue-1516684143) above to make it more visible. Please lemme know if you think this would be helpful?



> > @danaschwanden Thanks for clarifying. Alternate Suggestion (if it makes sense) -> Do you mind calling out in the PR body the list of files specific to NIND feature that needs review and add a note that the other files have been modified to pass the linter and have no actual content changes. Thanks
> 
> Thanks a lot @vjayaramrh , had a quick exchange with @s3wong on slack earlier today and shared the following high level assessment with him:
> 
>     * 4 out of the 35 files are outside the NIND context (not tested by me so far) and updated to keep the linter happy
> 
>     * 31 files are related to NIND (tested implicitly by running NIND), of which 8 are net new (incl Dockerfile and pictures, etc)
> 
>     * Most of the linter driven changes are removing whitespaces in curly brackets which the linter seems to particularly dislike
> 
> 
> Long story short: The 4 files outside the NIND context boil down to the following:
> 
>     * 2 TF files (infra/compute_instances.tf and infra/output.tf) would benefit from some testing. Reckon running the TF code will quickly tell whether it is all good or something has been broken.
> 
>     * The other two files (from the ansible_kind folder) appear to be fairly benign changes as they are mostly removing spaces.
> 
> 
> Can add that to the [first comment](https://github.com/nephio-project/one-summit-22-workshop/pull/113#issue-1516684143) above to make it more visible. Please lemme know if you think this would be helpful?

@danaschwanden appreciate the response and the details you have provided in the PR body. It is very helpful.

File: nephio-ansible-install/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+RUN wget -q https://open-vsx.org/api/redhat/vscode-yaml/1.10.1/file/redhat.vscode-yaml-1.10.1.vsix && \
+unzip redhat.vscode-yaml-1.10.1.vsix "extension/*" && \
+mv extension /opt/code-oss/extensions/redhat.vscode-yaml
+
+RUN wget -q https://open-vsx.org/api/ms-kubernetes-tools/vscode-kubernetes-tools/1.3.11/file/ms-kubernetes-tools.vscode-kubernetes-tools-1.3.11.vsix && \
+unzip ms-kubernetes-tools.vscode-kubernetes-tools-1.3.11.vsix "extension/*" && \
+mv extension /opt/code-oss/extensions/ms.kubernetes-tools
+
+RUN mkdir -p /nephio-installation/inventory
+COPY ansible.cfg /nephio-installation/
```
Comment (vjayaramrh) [inline nephio-ansible-install/Dockerfile:20]: Is there a reason why the nind related files are in the nephio-ansible-install folder instead of being in its own folder (i.e sibling to nephio-ansible-install directory)? Thanks

**Comment by vjayaramrh** [comment]
@johnbelamaric Would it make sense to create a tag prior to this PR getting merged that has all the working changes till December?

File: nephio-ansible-install/NIND.md (hunk 1 above)
Comment (henderiw) [reply nephio-ansible-install/NIND.md:37]: it would be the same. I tested with a 4vcpu/16G as well, so this worked for me. now the more workloads we add not sure if this will still hold or not.

Comment (henderiw) [reply nephio-ansible-install/NIND.md:37]: so lets stay with the 32G, 8vcpu and we also have to specify the disk of 64G or so

File: nephio-ansible-install/NIND.md (hunk 2 above)
Comment (henderiw) [reply nephio-ansible-install/NIND.md:13]: Do we know if there is a pod man in padman option, as this uses docker in docker?

Comment (danaschwanden) [reply nephio-ansible-install/NIND.md:13]: +1
however, don't know about podman to comment with insights and agree it probably is best to cover that as part of a different PR (solicited via an Github issue as you @vjayaramrh suggested)

File: nephio-ansible-install/Dockerfile (hunk 3 above)
Comment (danaschwanden) [reply nephio-ansible-install/Dockerfile:20]: Quite often Dockerfiles are sitting in the same folder as the application. Reckoned it made sense to surface it there also for nind. Could move the assets to their own folder in case you think it makes sense?

File: nephio-ansible-install/NIND.md (hunk 1 above)
Comment (danaschwanden) [reply nephio-ansible-install/NIND.md:37]: The next sentence in the NIND.md file actually refers back to the main README.md with the workshop VM spec.
As @henderiw mention nind also runs with 4vcpu/16G (which i most often used during the develpment).
However, would also advocate to leave head room as the resource demands will only grow as the implementation progresses.

File: nephio-ansible-install/Dockerfile (hunk 3 above)
Comment (vjayaramrh) [reply nephio-ansible-install/Dockerfile:20]: I have seen go projects using `/build/package` directory or similar for Dockerfiles

File: nephio-ansible-install/NIND.md (hunk 2 above)
Comment (vjayaramrh) [reply nephio-ansible-install/NIND.md:13]: One can still run images built using Dockerfile with podman installed. For now, it makes sense to open an issue that tracks updates to be added to NIND.md with podman commands equivalent of the a) docker build b)docker run and c) docker exec

Comment (danaschwanden) [reply nephio-ansible-install/NIND.md:13]: Sounds good. Would you reckon this could fall also under your [issue 108](https://github.com/nephio-project/one-summit-22-workshop/issues/108) (_"add support in ansible playbooks to install podman"_) that you already raised (ie not just installing but also showing how it is run)?

Comment (vjayaramrh) [reply nephio-ansible-install/NIND.md:13]: The issue was opened last month to track only install of podman on VM. 
For NIND, I am guessing, the instructions for building the image, running the container are on a host (VM, BareMetal, workstation, etc) with docker already installed. If the host has podman installed of docker, we would want similar podman instructions. I believe they are separate issues in my opinion.

Comment (danaschwanden) [reply nephio-ansible-install/NIND.md:13]: Thanks @vjayaramrh , I added [issue 122](https://github.com/nephio-project/one-summit-22-workshop/issues/122) as a placeholder for the idea of implementing a similar concept with podman.

File: nephio-ansible-install/Dockerfile (hunk 3 above)
Comment (danaschwanden) [reply nephio-ansible-install/Dockerfile:20]: Good point and see your point @vjayaramrh
IMHO in our case I think the Dockerfile is in a suitable location in the nephio-ansible-install folder.
Like this the folder could serve as a multiple purpose one stop shop depending on how an implementer would like to install nephio.

1. Installing nephio by by running the ansible playbooks on a suitable VM (as exercised in the workshop)
2. In a complete end-to-end integrated containerized fashion with nind (by simply building and executing the container specified in the Dockerfile)
3. Somewhat/somewhere in between installing nephio by runnning the ansible playbooks in a suitable (base)container image (this is still work in progress but @henderiw and i discuss on a follow up PR for this)

This would give an implementer maximum flexibility on how they choose to explore nephio.

Hence, long story short I think the Dockerfile is well positioned in this folder as it fits in well with the storyline above and also makes it easily discoverable.
Apologies for the wall of words but thought it might help outline the reasoning behind the choice.
